deepwell/src/services/email/service.rs

```diff
@@ -0,0 +1,95 @@
+ * Copyright (C) 2019-2023 Wikijump Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use super::prelude::*;
+
+#[derive(Debug)]
+pub struct EmailService;
+
+impl EmailService {
+    /// Validates an email through the MailCheck API.
+    pub async fn validate(email: &String) -> Result<EmailValidationOutput> {
+        // Sends a GET request to the MailCheck API and deserializes the response.
+        let mailcheck = surf::get(format!("https://api.mailcheck.ai/email/{email}"))
+            .send()
+            .await?
+            .body_json::<MailCheckResponse>()
+            .await?;
+
+        // Create the output with default parameters.
+        let mut output = EmailValidationOutput::default();
+
+        // Check request status.
+        match mailcheck.status {
+            // Valid request.
+            200 => {}
+
+            // Invalid request.
+            400 => {
+                tide::log::error!(
+                    "MailCheck API returned status 400: {:?}",
+                    mailcheck.error
+                );
+                return Err(Error::BadRequest);
+            }
+
+            // Exceeded rate limit.
+            429 => {
+                tide::log::error!(
+                    "MailCheck API returned status 429: {:?}",
+                    mailcheck.error
+                );
+                return Err(Error::RateLimited);
+            }
+
+            // Other statuses.
+            _ => {
+                tide::log::warn!(
+                    "MailCheck API returned status {}: {:?}",
+                    mailcheck.status,
+                    mailcheck.error
+                );
+            }
+        }
+
+        // Check if the email is an alias.
+        if mailcheck.alias {
+            output.classification = EmailClassification::Alias;
+        }
+
+        // Check if the email is a disposable.
+        if mailcheck.disposable {
+            output.valid = false;
+            output.classification = EmailClassification::Disposable;
+        }
+
+        // Check if the domain has any MX records.
+        if !mailcheck.mx {
+            output.valid = false;
+            output.classification = EmailClassification::Invalid;
+        }
+
+        if mailcheck.did_you_mean.is_some() {
+            output.did_you_mean = mailcheck.did_you_mean;
+        }
```

Comment:
Normally wouldn't this be `output.did_you_mean = mailcheck.did_you_mean` unconditionally if we're copying `Option` for `Option` anyways? Otherwise it seems like a more canonical conditional would be
```rust
if let Some(suggestion) = mailcheck.did_you_mean {
    output.did_you_mean = Some(suggestion);
}
```
?

Between the two I'd rather do unconditional copy since that avoids a branch altogether.